### PR TITLE
[workflow] Add a 'veristat' job

### PR DESCRIPTION
**Pull request by eddyz87**
Adds a job to execute veristat on a selected set of object files as a part of the general test pipeline. The job is executed differently for pushes and for pull requests:

- For pushes:
  - the kernel is compiled as usual;
  - selftests are compiled as well, this is necessary to produce object files that would be used by veristat;
  - veristat is executed inside qemu, performance statistics is collected and cached as baseline data.

- For pull requests:
  - veristat is executed inside qemu and performance statistics is collected;
  - baseline data is restored from cache;
  - baseline data is compared to current statistics and step summary page is formed;
  - if there are some programs that verified successfully in the baseline but fail verification in current run, the step is marked as failed.

This also adds 'run_veristat' "entrypoint" for run_selftests.sh.

This entrypoint executes veristat on a set of binary files specified in `./tools/testing/selftests/bpf/veristat.cfg` file and saves output to the `/command_output/` directory within qemu image. This output is extracted from the image by `run-qemu` action.

This pull request depends on the following pull requests for libbpf/ci repository:
- https://github.com/libbpf/ci/pull/98
- https://github.com/libbpf/ci/pull/99

I tested these changes using my own fork of the https://github.com/kernel-patches/bpf repository.
For test purposes I had to do some adjustments compared to production:
- minimize set of tests;
- change 'runs_on' actions to execution on my machine;
- link to eddyz87/libbpf-ci instead of eddyz87/libbpf/ci where appropriate.

Here are example runs:
- CI run for push to 'bpf-next':
  https://github.com/eddyz87/bpf/actions/runs/4692697761

- CI run for feature branch that breaks verification for some of the files
  (note the 'Summary' page, same info is duplicated in the text log for the `veristat` job):
  https://github.com/eddyz87/bpf/actions/runs/4692808046

- CI run for feature that does not change verification results:
  https://github.com/eddyz87/bpf/actions/runs/4692963478

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -353,3 +355,84 @@ jobs:
...
+          set -e
+          cd ${{ github.workspace }}
+          selftests/bpf/veristat -o csv -e file,prog,verdict,states -C \
+            veristat-baseline.csv veristat.csv > compare.csv
+          python3 ./.github/scripts/veristat-compare.py compare.csv
+
+      # For push: just put baseline log to cache
+      - if: ${{ github.event_name == 'push' }}
+        run: |
+          cp '${{ github.workspace }}/veristat.csv' \
```
Comment (danielocfb) [inline .github/workflows/test.yml:431]: Nit: Can this be a `mv`, actually?

File: ci/vmtest/run_selftests.sh [hunk 2]
```diff
@@ -98,6 +99,17 @@ test_verifier() {
   foldable end test_verifier
 }
 
+run_veristat() {
+  foldable start run_veristat "Running veristat"
+
+  globs=$(awk '/^#/ { next; } { print $0 ".bpf.o"; }' ./veristat.cfg)
+  mkdir -p ${OUTPUT_DIR}
+  ./veristat -o csv -q -e file,prog,verdict,states ${globs} > ${OUTPUT_DIR}/veristat.csv
```
Comment (danielocfb) [inline ci/vmtest/run_selftests.sh:107]: Can we please use long options, if available?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -353,3 +355,84 @@ jobs:
...
+        with:
+          key: x86_64-gcc-veristat-baseline
+          restore-keys: |
+            x86_64-gcc-veristat-baseline-
+          path: '${{ github.workspace }}/veristat-baseline.csv'
+
+      - if: ${{ github.event_name == 'pull_request' }}
+        name: Show veristat comparison
+        run: |
+          set -e
```
Comment (danielocfb) [inline .github/workflows/test.yml:422]: Unnecessary? https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -295,16 +299,14 @@ jobs:
             --exclude '*.h' \
             --exclude '*.output' \
             selftests/bpf/ | zstd -T0 -19 -o vmlinux-${{ matrix.arch }}-${{ matrix.toolchain_full }}.tar.zst
-      - if: ${{ github.event_name != 'push' }}
-        name: Remove KBUILD_OUTPUT contents
+      - name: Remove KBUILD_OUTPUT contents
```
Comment (danielocfb) [inline .github/workflows/test.yml:302]: Can you please clarify why this change is necessary? It strikes me as wrong. Now we are no longer creating caches at all, right? I.e., it breaks incremental builds.
Also, comment in line 305 is outdated now.

Comment (eddyz87) [reply .github/workflows/test.yml:302]: This might be unnecessary indeed. I think I read it as '==' for some reason.
Thanks for all the comments, I'll reply in each after making changes and re-testing.

File: .github/scripts/veristat-compare.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    print_summary()
+    print_summary('<details>')
+    print_summary('<summary>Click to expand</summary>')
+    print_summary()
+    print_table(html_table, print_summary)
+    print_summary()
+    print_summary('</details>')
+
+if new_failures:
+    print(f'Failing step because of the new failures')
```
[the remark this answers is not in the document]
Comment (eddyz87) [reply .github/scripts/veristat-compare.py]: Removed

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -353,3 +355,84 @@ jobs:
...
+    env:
+      KERNEL: LATEST
+      REPO_ROOT: ${{ github.workspace }}
+      REPO_PATH: ""
+      KBUILD_OUTPUT: kbuild-output/
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/download-artifact@v3
+        with:
+          name: vmlinux-x86_64-gcc
```
[the remark this answers is not in the document]
Comment (eddyz87) [reply .github/workflows/test.yml]: Added to env variables at the top level:

```yaml
env:
  veristat_arch: x86_64
  veristat_toolchain: gcc
```

And a step to derive the prefix at the start of the veristat job:

```yaml
      - name: Setup environment variables
        run: |
            echo arch_and_tool=${{ env.veristat_arch }}-${{ env.veristat_toolchain }} > \
              ${GITHUB_ENV}
```

Please let me know if there is a more elegant way to do it, extending `env` section at the job level to introduce the var using top-level `env.veristat_*` values didn't work for some reason.

Also changed `set-matrix` code that computes `veristat_runs_on` to use the above variables.

File: .github/workflows/test.yml (hunk 1 above)
Comment (eddyz87) [reply .github/workflows/test.yml]: Changed to `mv`.

File: ci/vmtest/run_selftests.sh (hunk 2 above)
Comment (eddyz87) [reply ci/vmtest/run_selftests.sh:107]: Done

File: .github/workflows/test.yml (hunk 3 above)
Comment (eddyz87) [reply .github/workflows/test.yml]: Unnecessary indeed, removed.

File: .github/workflows/test.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (danielocfb) [reply .github/workflows/test.yml]: What you do seems fine, thanks!

**Review by danielocfb — APPROVED** [review]
Looks good to me now, thanks! Will merge tomorrow, unless Manu wants to take a look as well.

**Comment by danielocfb** [comment]
We are seeing failures in CI, e.g., https://github.com/kernel-patches/bpf/actions/runs/4756853058/jobs/8453383397

Will revert the change for the time being.

**Comment by eddyz87** [comment]
It complains that veristat baseline is not present. The base line is generated after CI execution for a push to bpf-next. So, it won't be available for some time. Is this pipeline for pull request of a push? I can relax the condition, allowing this step to not fail if there is no baseline.

**Comment by danielocfb** [comment]
Correct, yes, we should not error out if it's not present, I would say. In general we expect to have a baseline available (just by virtue of how frequently we see pushes), but we cannot guarantee that. E.g., we've had long stretches in the past where no changes got pushed and it's conceivable that GitHub Actions evicts cached artifacts (we have limited control over that).

**Comment by danielocfb** [comment]
> Is this pipeline for pull request of a push?

This is the job for the push: https://github.com/kernel-patches/bpf/actions/runs/4756852477/jobs/8453428926
It did succeed. But let's fix this up.

**Comment by eddyz87** [comment]
Ok, will do that. I need to create new merge request for that, right?
I'll post an update in ~2hr, need to wrap up another task + local testing takes some time.

**Comment by danielocfb** [comment]
> Ok, will do that. I need to create new merge request for that, right?

Yeah, please do that. Will probably get merged only next week then, because I'll be out tomorrow.
